cpu/sam0_common/periph/i2c.c

```diff
@@ -329,48 +421,117 @@ void _i2c_poweroff(i2c_t dev)
 static int _i2c_start(SercomI2cm *dev, uint16_t addr)
 {
     /* Wait for hardware module to sync */
-    DEBUG("Wait for device to be ready\n");
+    DEBUG_PUTS("i2c.c: Wait for device to be ready");
     _syncbusy(dev);
 
     /* Set action to ACK. */
     dev->CTRLB.reg &= ~SERCOM_I2CM_CTRLB_ACKACT;
 
     /* Send Start | Address | Write/Read */
-    DEBUG("Generate start condition by sending address\n");
+    DEBUG("i2c.c: Generate start condition by sending address 0x%04"PRIu16"\n",
+            addr);
     dev->ADDR.reg = addr;
 
-    /* Wait for response on bus. */
-    if (addr & I2C_READ) {
-        /* Some devices (e.g. SHT2x) can hold the bus while
-        preparing the reply */
-        if (_wait_for_response(dev, 100 * SAMD21_I2C_TIMEOUT) < 0)
-            return -ETIMEDOUT;
-    }
-    else {
-        if (_wait_for_response(dev, SAMD21_I2C_TIMEOUT) < 0)
-            return -ETIMEDOUT;
-    }
+    /* Wait for response on bus.
+     * Some devices (e.g. SHT2x) can hold the bus while
+     * preparing the reply. */
+    uint32_t timeout = (addr & I2C_READ) ? 100 * SAMD21_I2C_TIMEOUT : SAMD21_I2C_TIMEOUT;
+    int res = _wait_for_response(dev, timeout);
+
+    /* According to section 36.6.2.4.2 of the SAMD51/SAME54 datasheet
+     * (also confirmed for SAML21),
+     * the following flags are meaningful in address transmission
+     * and will be checked here:
+     *
+     * - STATUS.ARBLOST: Arbitration lost, which should be
+     *   considered when INTFLAG.MB is set.
+     *   STATUS.BUSERR should also be set in this case
+     *   according to the datasheet.
+     * - STATUS.RXNACK: ACK not received after sending the address.
+     *   Addressed device is busy or not present. This will also be set
+     *   with INTFLAG.MB.
+     *
+     * We further generally check for BUSERR and unexpected bus states.
+     *
+     * We ignore dev->INTFLAG.bit.ERROR altogehter as it does not
+     * seem to get set in the current configuration of the SERCOM module
+     * (this finding applies to SAME54 / SAMD51 devices).
+     *
+     * We ignore the three timeout flags in the status register:
+     * SERCOM_I2CM_STATUS_MEXTTOUT, SERCOM_I2CM_STATUS_LOWTOUT
+     * SERCOM_I2CM_STATUS_SEXTTOUT.
+     *
+     * We further ignore SERCOM_I2CM_STATUS_LENERR (useful in 32 bit
+     * mode and with DMA, not related to address transmission).
+     */
+    uint16_t intflag = dev->INTFLAG.reg;
+    uint16_t status = dev->STATUS.reg;
+    bool any_error = res ||
+        (
+            (intflag & SERCOM_I2CM_INTFLAG_MB) &&
+            (status & (SERCOM_I2CM_STATUS_ARBLOST | SERCOM_I2CM_STATUS_RXNACK))
+        ) || (
+            status & SERCOM_I2CM_STATUS_BUSERR
+        ) || (
+            (status & SERCOM_I2CM_STATUS_BUSSTATE_Msk) == BUSSTATE_UNKNOWN
+        ) || (
+            (status & SERCOM_I2CM_STATUS_BUSSTATE_Msk) == BUSSTATE_BUSY
+        );
```

Comment:
Why do you need this when you check all those conditions individually afterwards anyway?

Reply:
I tried to count the instructions off in compiler explorer. In the case that there is not error, all if clauses are executed and all parts of the compound boolean expression are evaluated as well. It seems like there is no advantage on using this boolean expression to test for any error. I will therefore remove this part as you suggest.

Reply:
changed